src/python/pants/backend/experimental/python/publish_pypi.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.target_types import ConsoleScript, PythonDistribution
+from pants.backend.python.util_rules.pex import PexRequest, PexRequirements, VenvPex, VenvPexProcess
+from pants.core.goals.publish import (
+    PublishProcess,
+    PublishRequest,
+    PublishTarget,
+    PublishTargetField,
+)
+from pants.engine.environment import Environment, InterpolatedEnvironmentRequest
+from pants.engine.process import InteractiveProcess, Process
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import StringField
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+class Twine(PythonToolBase):
+    options_scope = "twine"
+    help = "The utility for publishing Python pakcages to PyPi and other repositories."
+    default_version = "twine==3.4.2"
+    default_main = ConsoleScript("twine")
+    register_interpreter_constraints = True
+    default_interpreter_constraints = ["CPython>=3.6"]
+
+
+class PypiRepositoryName(StringField):
+    alias = "repository"
+    help = "The pypi repository to upload the package to."
+    default = "pypi"
+
+
+class PypiRepositoryUrl(StringField):
+    alias = "repository_url"
+    help = "The pypi repository url to upload the package to. Takes priority over `repository`."
+
+
+class PypiRepositoryUsername(StringField):
+    alias = "username"
+    help = "The username for this repository. Suppirting ${BASH_STYLE} environment interpolation."
+    default = "${TWINE_USERNAME}"
+
+
+class PypiRepositoryPassword(StringField):
+    alias = "password"
+    help = "The password for this repository. Supporting ${BASH_STYLE} environment interpolation."
+    default = "${TWINE_PASSWORD}"
```

Comment:
@Eric-Arellano : Is this syntax something that seems viable as a new standard?

Reply:
Yeah, I think so. I think it would fit well in pants.toml too for generic env var support. 

The even safer alternative is to make this always be an env var name without needing escaping...should users ever set the password as a literal value? That seems like setting our users up for trouble. Field could be password_env_var

Reply:
I think I prefer things like this to live in config files, rather than in BUILD files.. (this whole file is from Chris, so I'm picking it apart into the new `pants.backend.python.goals.publish` which is not in its final form yet either.)

So, my plan was to get rid of the `pypi_repository` as a build target, and instead have it be a subsystem, like I did for docker registries. wdyt?

However, env var support for config values is a natural thing to do, imo. You shouldn't hard code secrets any where, ever.

Reply:
> So, my plan was to get rid of the `pypi_repository` as a build target, and instead have it be a subsystem, like I did for docker registries. wdyt?

This is probably fine. The primary challenge just becomes supporting multiple of them. We don't have a great razor for "what should go in targets" vs "what should go in config", but one potentially useful angle is: how many of them will there be?

If it's possible that many different projects in a repository will need to be published to many different places, then targets might be a better pattern. On the other hand, if the whole repository is likely to only publish to a handful of destinations (common for most language and docker packages probably... the most common number of destinations is probably 2: one for OSS, and one for closed source), then config is probably fine... although as soon as the number of destinations is >1, you need to have a structure in config that supports naming the destinations.

Reply:
In this case, the repositories are configured in `.pypirc` files, and I feel that is the right thing to do, as then it works the same if you'd want to run `twine` yourself too.